File: services/booking/internal/domain/booking.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// Status represents the lifecycle state of a booking.
type Status string

const (
	StatusPendingSeat         Status = "pending_seat"
	StatusPendingPayment      Status = "pending_payment"
	StatusConfirmed           Status = "confirmed"
	StatusCancelled           Status = "cancelled"
	StatusExpired             Status = "expired"
	StatusFailed              Status = "failed"
	StatusRefunded            Status = "refunded"
	StatusUsed                Status = "used"
	StatusDisputed            Status = "disputed"
	StatusPartiallyCancelled  Status = "partially_cancelled"
	StatusPartiallyRefunded   Status = "partially_refunded"
)

// PassengerCategory drives discount calculation.
type PassengerCategory string

const (
	CategoryAdult   PassengerCategory = "adult"
	CategoryChild   PassengerCategory = "child"
	CategorySenior  PassengerCategory = "senior"
	CategoryStudent PassengerCategory = "student"
)

// Trip is the value-object snapshot of trip data needed at booking time.
type Trip struct {
	ID             uuid.UUID
	OperatorID     uuid.UUID
	PriceCents     int
	Currency       string
	AvailableSeats int
	TotalSeats     int
	BusClass       string
}

// Passenger holds the traveller details for one seat on a booking.
type Passenger struct {
	FirstName string
	LastName  string
	Category  PassengerCategory
}

// BookingSeat is one seat-passenger pairing inside a booking.
type BookingSeat struct {
	ID         uuid.UUID
	BookingID  uuid.UUID
	SeatNumber string
	Passenger  Passenger
	PriceCents int
}

// Booking is the aggregate root for a reservation.
type Booking struct {
	ID              uuid.UUID
	UserID          uuid.UUID
	TripID          uuid.UUID
	Status          Status
	TotalPriceCents int
	Currency        string
	LockExpiresAt   time.Time
	ConfirmedAt     *time.Time
	CancelledAt     *time.Time
	CreatedAt       time.Time
	UpdatedAt       time.Time
	Seats           []BookingSeat
}

File: web/src/types/shared/booking.ts
export enum BookingStatus {
  PENDING_SEAT = "PENDING_SEAT",
  PENDING_PAYMENT = "PENDING_PAYMENT",
  CONFIRMED = "CONFIRMED",
  CANCELLED = "CANCELLED",
  PARTIALLY_CANCELLED = "PARTIALLY_CANCELLED",
  REFUNDED = "REFUNDED",
  PARTIALLY_REFUNDED = "PARTIALLY_REFUNDED",
  EXPIRED = "EXPIRED",
  USED = "USED",
  FAILED = "FAILED",
  DISPUTED = "DISPUTED",
}

export type PassengerCategory = "adult" | "child" | "senior" | "student";

export interface BookingPassenger {
  readonly id: string;
  readonly firstName: string;
  readonly lastName: string;
  readonly category: PassengerCategory;
  readonly seatId: string;
  readonly unitPrice: number;
  readonly baggageKg: number;
  readonly baggageFee: number;
}

export interface Booking {
  readonly id: string;
  readonly userId?: string;
  readonly tripId: string;
  readonly passengers: readonly BookingPassenger[];
  readonly totalAmount: number;
  readonly currency: string;
  readonly status: BookingStatus;
  readonly idempotencyKey: string;
  readonly createdAt: string;
  readonly updatedAt: string;
  readonly expiresAt?: string;
}
